node/network/collator-protocol/src/collator_side/validators_buffer.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright 2017-2022 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Validator groups buffer for connection managements.
+//!
+//! Solves 2 problems:
+//! 	1. A collator may want to stay connected to multiple groups on rotation boundaries.
+//! 	2. It's important to disconnect from validator when there're no collations to be fetched.
+//!
+//! We keep a simple FIFO buffer of N validator groups and a bitvec for each advertisement,
+//! 1 indicating we want to be connected to i-th validator in a buffer, 0 otherwise.
+//!
+//! The bit is set to 1 on new advertisements, and back to 0 when a collation is fetched
+//! by a validator.
+//!
+//! The bitwise OR over known advertisements gives us validators indices for connection request.
+
+use std::{
+	collections::{HashMap, VecDeque},
+	ops::Range,
+};
+
+use bitvec::{bitvec, vec::BitVec};
+
+use polkadot_primitives::v2::{AuthorityDiscoveryId, GroupIndex, Hash, SessionIndex};
+
+pub const VALIDATORS_BUFFER_CAPACITY: usize = 3;
+
+/// Unique identifier of a validators group.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+struct ValidatorsGroupInfo {
+	len: usize,
+	session_index: SessionIndex,
+	group_index: GroupIndex,
+}
+
+#[derive(Debug)]
+pub struct ValidatorGroupsBuffer<const N: usize> {
+	buf: VecDeque<ValidatorsGroupInfo>,
+	validators: VecDeque<AuthorityDiscoveryId>,
+	should_be_connected: HashMap<Hash, BitVec>,
+}
+
+impl<const N: usize> ValidatorGroupsBuffer<N> {
+	pub fn new() -> Self {
+		assert!(N > 0);
```

Comment:
panics in production code which are not provably impossible aren't allowed

Reply:
70bffc7365777089f6f84af524cb0c2938cbeddb